uniffi/src/scaffolding.rs

```diff
@@ -42,41 +39,38 @@ mod filters {
 
     pub fn ret_type_rs(type_: &TypeReference) -> Result<String, askama::Error> {
         Ok(match type_ {
-            // These can be passed directly over the FFI without conversion.
             TypeReference::U32 => "u32".to_string(),
             TypeReference::U64 => "u64".to_string(),
             TypeReference::Float => "f32".to_string(),
             TypeReference::Double => "f64".to_string(),
-            TypeReference::Boolean => "u8".to_string(),
-            // These types need conversion, and will require special handling below
-            // when lifting/lowering.
+            TypeReference::Boolean => "bool".to_string(),
             TypeReference::String => "String".to_string(),
             TypeReference::Enum(name) => name.clone(),
             TypeReference::Record(name) => name.clone(),
             TypeReference::Optional(t) => format!("Option<{}>", type_rs(t)?),
-            _ => panic!("[TODO: type_rs({:?})]", type_),
+            _ => panic!("[TODO: ret_type_rs({:?})]", type_),
         })
     }
 
     pub fn type_c(type_: &TypeReference) -> Result<String, askama::Error> {
         Ok(match type_ {
-            TypeReference::String => "ffi_support::FfiStr<'_>".to_string(),
-            TypeReference::Enum(_) => "u32".to_string(),
-            TypeReference::Record(_) => "ffi_support::ByteBuffer".to_string(),
-            TypeReference::Optional(_) => "ffi_support::ByteBuffer".to_string(),
+            // Objects don't currently impl `ViaFfi`.
             TypeReference::Object(_) => "u64".to_string(),
-            _ => type_rs(type_)?,
+            _ => format!(
+              "<{} as uniffi::support::ViaFfi>::Value",
+              type_rs(type_)?,
+            ),
```

Comment:
I discovered that you can use the associated types of a trait in the type signature of a function. Expressing it like this means we don't have to repeat the big list of "rust type X maps to FFI type Y" here when generating the scaffolding, we can just look up the mapping on the `ViaFfi` trait.

I like it, but some downsides of this approach:

* The generating rust scaffolding is a bit more visually complex; I would argue that it's actually easier to *understand* because you don't have to account for the mapping between Rust and C types, but it's definitely extra syntax noise.
* This only works if the corresponding rust types are public, since they appear in the type signature of a public function. Hence all the additions of `pub` in the examples. But TBH I think it's reasonable to insist that these types be marked `pub` because they are part of your public API.

If you don't like this, I can revert the change can go back to a big `match` statement like we had here before, but I figured I'd try it out.

Reply:
This is the kind of magic that I like, I'll take 1 source of truth over 3, nice cleanup :)